README.rst
Python Client for Google Cloud Dataproc API
===========================================

|stable| |pypi| |versions|

`Google Cloud Dataproc API`_: 

- `Client Library Documentation`_
- `Product Documentation`_

.. |stable| image:: https://img.shields.io/badge/support-stable-gold.svg
   :target: https://github.com/googleapis/google-cloud-python/blob/main/README.rst#stability-levels
.. |pypi| image:: https://img.shields.io/pypi/v/google-cloud-dataproc.svg
   :target: https://pypi.org/project/google-cloud-dataproc/
.. |versions| image:: https://img.shields.io/pypi/pyversions/google-cloud-dataproc.svg
   :target: https://pypi.org/project/google-cloud-dataproc/
.. _Google Cloud Dataproc API: https://cloud.google.com/dataproc
.. _Client Library Documentation: https://cloud.google.com/python/docs/reference/dataproc/latest
.. _Product Documentation:  https://cloud.google.com/dataproc

Quick Start
-----------

In order to use this library, you first need to go through the following steps:

1. `Select or create a Cloud Platform project.`_
2. `Enable billing for your project.`_
3. `Enable the Google Cloud Dataproc API.`_
4. `Setup Authentication.`_

.. _Select or create a Cloud Platform project.: https://console.cloud.google.com/project
.. _Enable billing for your project.: https://cloud.google.com/billing/docs/how-to/modify-project#enable_billing_for_a_project
.. _Enable the Google Cloud Dataproc API.:  https://cloud.google.com/dataproc
.. _Setup Authentication.: https://googleapis.dev/python/google-api-core/latest/auth.html

Installation
~~~~~~~~~~~~

Install this library in a `virtualenv`_ using pip. `virtualenv`_ is a tool to
create isolated Python environments. The basic problem it addresses is one of
dependencies and versions, and indirectly permissions.

With `virtualenv`_, it's possible to install this library without needing system
install permissions, and without clashing with the installed system
dependencies.

.. _`virtualenv`: https://virtualenv.pypa.io/en/latest/


Code samples and snippets
~~~~~~~~~~~~~~~~~~~~~~~~~

Code samples and snippets live in the `samples/` folder.


Supported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^
Our client libraries are compatible with all current [active](https://devguide.python.org/devcycle/#in-development-main-branch) and [maintenance](https://devguide.python.org/devcycle/#maintenance-branches) versions of
Python.

Python >= 3.7

Unsupported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^^^
Python <= 3.6

If you are using an [end-of-life](https://devguide.python.org/devcycle/#end-of-life-branches)
version of Python, we recommend that you update as soon as possible to an actively supported version.


Mac/Linux
^^^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    source <your-env>/bin/activate
    <your-env>/bin/pip install google-cloud-dataproc


Windows
^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    <your-env>\Scripts\activate
    <your-env>\Scripts\pip.exe install google-cloud-dataproc

Next Steps
~~~~~~~~~~

-  Read the `Client Library Documentation`_ for Google Cloud Dataproc API
   to see other available methods on the client.
-  Read the `Google Cloud Dataproc API Product documentation`_ to learn
   more about the product and see How-to Guides.
-  View this `README`_ to see the full list of Cloud
   APIs that we cover.

.. _Google Cloud Dataproc API Product documentation:  https://cloud.google.com/dataproc
.. _README: https://github.com/googleapis/google-cloud-python/blob/main/README.rst


CHANGELOG.md
# Changelog

[PyPI History][1]

[1]: https://pypi.org/project/google-cloud-dataproc/#history

## [5.1.0](https://github.com/googleapis/python-dataproc/compare/v5.0.1...v5.1.0) (2022-09-27)


### Features

* Add support for Dataproc metric configuration ([#481](https://github.com/googleapis/python-dataproc/issues/481)) ([6d0e404](https://github.com/googleapis/python-dataproc/commit/6d0e404927e9af4f7f8b07ca76604a22cc57b787))

## [5.0.1](https://github.com/googleapis/python-dataproc/compare/v5.0.0...v5.0.1) (2022-08-11)


### Bug Fixes

* **deps:** allow protobuf < 5.0.0 ([#456](https://github.com/googleapis/python-dataproc/issues/456)) ([a446937](https://github.com/googleapis/python-dataproc/commit/a44693711df3218a083f060e00cad3801537dd9b))
* **deps:** require proto-plus >= 1.22.0 ([a446937](https://github.com/googleapis/python-dataproc/commit/a44693711df3218a083f060e00cad3801537dd9b))

## [5.0.0](https://github.com/googleapis/python-dataproc/compare/v4.0.3...v5.0.0) (2022-07-19)


### ⚠ BREAKING CHANGES

* Move `yarn_config` into a `oneof`
* Remove `temp_bucket` from VirtualClusterConfig, as its value was not used

### Features

* add audience parameter ([61a23fa](https://github.com/googleapis/python-dataproc/commit/61a23faab861c17043af4efeeb1659334234349a))


### Bug Fixes

* **deps:** require google-api-core>=1.32.0,>=2.8.0 ([#400](https://github.com/googleapis/python-dataproc/issues/400)) ([61a23fa](https://github.com/googleapis/python-dataproc/commit/61a23faab861c17043af4efeeb1659334234349a))
* Move `yarn_config` into a `oneof` ([61a23fa](https://github.com/googleapis/python-dataproc/commit/61a23faab861c17043af4efeeb1659334234349a))
* Remove `temp_bucket` from VirtualClusterConfig, as its value was not used ([61a23fa](https://github.com/googleapis/python-dataproc/commit/61a23faab861c17043af4efeeb1659334234349a))
* require python 3.7+ ([#442](https://github.com/googleapis/python-dataproc/issues/442)) ([9862ff7](https://github.com/googleapis/python-dataproc/commit/9862ff7c9086921f0a4ecf81ae175c07ac701ef3))

## [4.0.3](https://github.com/googleapis/python-dataproc/compare/v4.0.2...v4.0.3) (2022-06-03)


### Bug Fixes

* **deps:** require protobuf <4.0.0dev ([#423](https://github.com/googleapis/python-dataproc/issues/423)) ([5d1a263](https://github.com/googleapis/python-dataproc/commit/5d1a263533c1812587a9668d6bd6d7d29ac82f2f))


### Documentation

* fix changelog header to consistent size ([#424](https://github.com/googleapis/python-dataproc/issues/424)) ([00162f0](https://github.com/googleapis/python-dataproc/commit/00162f07935cca365c41dcbf8be4e5a7681d680e))

## [4.0.2](https://github.com/googleapis/python-dataproc/compare/v4.0.1...v4.0.2) (2022-04-06)


### Bug Fixes

* resource quotas ([#377](https://github.com/googleapis/python-dataproc/issues/377)) ([122c2f7](https://github.com/googleapis/python-dataproc/commit/122c2f77a93228dda409a8fad22465f2d28c5e0d))
* updating submit_job_to_cluster.py ([#387](https://github.com/googleapis/python-dataproc/issues/387)) ([0719d2b](https://github.com/googleapis/python-dataproc/commit/0719d2b69661f9775c00a1fc0dade2e65b4e44e9))

## [4.0.1](https://github.com/googleapis/python-dataproc/compare/v4.0.0...v4.0.1) (2022-03-05)


### Bug Fixes

* **deps:** require google-api-core>=1.31.5, >=2.3.2 ([#357](https://github.com/googleapis/python-dataproc/issues/357)) ([3c66f42](https://github.com/googleapis/python-dataproc/commit/3c66f4232d35f1e11807d29e169fe87e028c52eb))
* **deps:** require proto-plus>=1.15.0 ([3c66f42](https://github.com/googleapis/python-dataproc/commit/3c66f4232d35f1e11807d29e169fe87e028c52eb))

## [4.0.0](https://github.com/googleapis/python-dataproc/compare/v3.3.0...v4.0.0) (2022-02-26)


### ⚠ BREAKING CHANGES

* add support for Virtual Dataproc cluster running on GKE cluster (#344)

### Features

* add support for Virtual Dataproc cluster running on GKE cluster ([#344](https://github.com/googleapis/python-dataproc/issues/344)) ([116077b](https://github.com/googleapis/python-dataproc/commit/116077b45abaccb1814002284e05e34ef387e045))


### Bug Fixes

* move GkeClusterConfig to google.cloud.dataproc_v1.types.shared ([116077b](https://github.com/googleapis/python-dataproc/commit/116077b45abaccb1814002284e05e34ef387e045))
* remove namespaced_gke_deployment_target ([116077b](https://github.com/googleapis/python-dataproc/commit/116077b45abaccb1814002284e05e34ef387e045))

## [3.3.0](https://github.com/googleapis/python-dataproc/compare/v3.2.0...v3.3.0) (2022-02-18)


### Features

* add api key support ([#336](https://github.com/googleapis/python-dataproc/issues/336)) ([ac22d7e](https://github.com/googleapis/python-dataproc/commit/ac22d7ef7040e85035a8d3cfc9fe0f69a014f238))


### Bug Fixes

* resolve DuplicateCredentialArgs error when using credentials_file ([452460f](https://github.com/googleapis/python-dataproc/commit/452460fe8e6af9e1a99d9636c5531e489e1e4852))


### Documentation

* add generated snippets ([#342](https://github.com/googleapis/python-dataproc/issues/342)) ([98810a9](https://github.com/googleapis/python-dataproc/commit/98810a9bc7674ea81397823f5da871cd30adcbd7))

## [3.2.0](https://github.com/googleapis/python-dataproc/compare/v3.1.1...v3.2.0) (2022-01-17)


### Features

* add Spark runtime versioning for Spark batches ([#318](https://github.com/googleapis/python-dataproc/issues/318)) ([f2e35d9](https://github.com/googleapis/python-dataproc/commit/f2e35d9735cbd0dd5a0e32d78631d70820380846))
* auto-diagnostic of failed Spark batches ([f2e35d9](https://github.com/googleapis/python-dataproc/commit/f2e35d9735cbd0dd5a0e32d78631d70820380846))
* custom image containers for Spark batches ([f2e35d9](https://github.com/googleapis/python-dataproc/commit/f2e35d9735cbd0dd5a0e32d78631d70820380846))
* local SSD NVME interface support for GCE clusters ([f2e35d9](https://github.com/googleapis/python-dataproc/commit/f2e35d9735cbd0dd5a0e32d78631d70820380846))


owl-bot-staging/v1/noxfile.py
# -*- coding: utf-8 -*-
# Copyright 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import os
import pathlib
import shutil
import subprocess
import sys


import nox  # type: ignore

ALL_PYTHON = [
    "3.6",
    "3.7",
    "3.8",
    "3.9",
    "3.10",
]

CURRENT_DIRECTORY = pathlib.Path(__file__).parent.absolute()

LOWER_BOUND_CONSTRAINTS_FILE = CURRENT_DIRECTORY / "constraints.txt"
PACKAGE_NAME = subprocess.check_output([sys.executable, "setup.py", "--name"], encoding="utf-8")

BLACK_VERSION = "black==22.3.0"
BLACK_PATHS = ["docs", "google", "tests", "samples", "noxfile.py", "setup.py"]
DEFAULT_PYTHON_VERSION = "3.10"

nox.sessions = [
    "unit",
    "cover",
    "mypy",
    "check_lower_bounds"
    # exclude update_lower_bounds from default
    "docs",
    "blacken",
    "lint",
    "lint_setup_py",
]

@nox.session(python=ALL_PYTHON)
def unit(session):
    """Run the unit test suite."""

    session.install('coverage', 'pytest', 'pytest-cov', 'pytest-asyncio', 'asyncmock; python_version < "3.8"')
    session.install('-e', '.')

    session.run(
        'py.test',
        '--quiet',
        '--cov=google/cloud/dataproc_v1/',
        '--cov=tests/',
        '--cov-config=.coveragerc',
        '--cov-report=term',
        '--cov-report=html',
        os.path.join('tests', 'unit', ''.join(session.posargs))
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def cover(session):
    """Run the final coverage report.
    This outputs the coverage report aggregating coverage from the unit
    test runs (not system test runs), and then erases coverage data.
    """
    session.install("coverage", "pytest-cov")
    session.run("coverage", "report", "--show-missing", "--fail-under=100")

    session.run("coverage", "erase")


@nox.session(python=ALL_PYTHON)
def mypy(session):
    """Run the type checker."""
    session.install('mypy', 'types-pkg_resources')
    session.install('.')
    session.run(
        'mypy',
        '--explicit-package-bases',
        'google',
    )


@nox.session
def update_lower_bounds(session):
    """Update lower bounds in constraints.txt to match setup.py"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'update',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )


@nox.session
def check_lower_bounds(session):
    """Check lower bounds in setup.py are reflected in constraints file"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'check',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )

@nox.session(python=DEFAULT_PYTHON_VERSION)
def docs(session):
    """Build the docs for this library."""

    session.install("-e", ".")
    session.install("sphinx==4.0.1", "alabaster", "recommonmark")

    shutil.rmtree(os.path.join("docs", "_build"), ignore_errors=True)
    session.run(
        "sphinx-build",
        "-W",  # warnings as errors
        "-T",  # show full traceback on exception
        "-N",  # no colors
        "-b",
        "html",
        "-d",
        os.path.join("docs", "_build", "doctrees", ""),
        os.path.join("docs", ""),
        os.path.join("docs", "_build", "html", ""),
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint(session):
    """Run linters.

    Returns a failure if the linters find linting errors or sufficiently
    serious code quality issues.
    """
    session.install("flake8", BLACK_VERSION)
    session.run(
        "black",
        "--check",
        *BLACK_PATHS,
    )
    session.run("flake8", "google", "tests", "samples")


@nox.session(python=DEFAULT_PYTHON_VERSION)
def blacken(session):
    """Run black. Format code to uniform standard."""
    session.install(BLACK_VERSION)
    session.run(
        "black",
        *BLACK_PATHS,
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint_setup_py(session):
    """Verify that setup.py is valid (including RST check)."""
    session.install("docutils", "pygments")
    session.run("python", "setup.py", "check", "--restructuredtext", "--strict")
